gslab_scons/tests/_test_helpers.py

```diff
@@ -0,0 +1,237 @@
+        misc_patch  = mock.patch(misc_path, platform)
+        total_patch = lambda f: misc_patch(main_patch(f))
+    except ImportError:
+        total_patch = main_patch
+
+    return total_patch
+
+def command_match(command, executable, which = None):
+    '''Parse Python, R, and Stata system calls as re.match objects'''
+    if executable in ['python', 'py']:
+        match = re.match('\s*'
+                         '(?P<executable>python)'
+                         '\s*'
+                         '(?P<source>[-\.\/\w]+)'
+                         '\s*'
+                         '(?P<args>(\s?[\.\/\w]+)*)?'
+                         '\s*'
+                         '(?P<log>>\s*[\.\/\w]+)?',
+                         command)
+
+    elif executable in ['r', 'R']:
+        match = re.match('\s*'
+                         '(?P<executable>R CMD BATCH)'
+                         '\s+'
+                         '(?P<option>--no-save)?'
+                         '\s*'
+                         "(?P<args>'--args .*')?"
+                         '\s*'
+                         '(?P<source>[\.\/\w]+\.[rR])'
+                         '\s*'
+                         '(?P<log>[\.\/\w]+(\.\w+)?)?',
+                         command)  
+
+    elif executable in ['stata', 'do']: 
+        match = re.match('\s*'
+                         '(?P<executable>\S+)'
+                         '\s+'
+                         '(?P<options>(\s?[-\/][-A-Za-z]+)+)?'
+                         '\s*'
+                         '(?P<do>do)?'
+                         '\s*'
+                         '(?P<source>[\.\/\w]+\.do)'
+                         '\s*'
+                         '(?P<args>.*)',
+                         command)
+
+    if which:
+        return match.group(which)
+    else:
+        return match
+
+
+def check_log(test_object, log_path, timestamp = True):
+    '''Check for the existence of a timestamped log'''
+    with open(log_path, 'rU') as log_file:
+        log_data = log_file.read()
+
+    if timestamp:
+        test_object.assertIn('Log created:', log_data)
+    else:
+        test_object.assertNotIn('Log created:', log_data)
+
+    os.remove(log_path)
+
+
+def bad_extension(test_object, builder, 
+                  bad  = 'bad',
+                  good = None,
+                  env  = {}):
+    '''Ensure builders fail when their first sources have bad extensions'''
+    if good:
+        source = [bad, good]
+    else:
+        source = [bad]
+
+    with test_object.assertRaises(BadExtensionError):
+        builder(target = './test_output.txt', 
+                source = [bad, good], 
+                env    = env)
+
+def standard_test(test_object, builder, 
+                  extension   = None,  
+                  system_mock = None,
+                  source = None,
+                  target = None,
+                  env    = None):
+    '''Test that builders run without errors and create logs properly.'''
+    if not source:
```

Comment:
I made a few changes along these lines in b9f18d8, but I kept a couple `source` definitions in function bodies because they require an `extension` argument.